core/graphelier-service/models/orderbooks.go

```diff
@@ -267,3 +267,24 @@ func (orderbook *Orderbook) ApplyMessagesToDeltabook(messages []*Message, numMes
 
 	return
 }
+
+// TopBookPerXNano : Returns a list of the top book at every x nanosecond after applying each individual message in a list
+func (orderbook *Orderbook) TopBookPerXNano(messages []*Message, x int) (topbook []*Point) {
+	firstNano := messages[0].Timestamp
+	lastNano := messages[len(messages)-x].Timestamp
+
+	for counter, nano := uint64(0), firstNano; nano <= lastNano; counter, nano = counter+1, nano+1 {
+		nanoMessages := GetMessagesForTimestamp(messages, nano)
+		orderbook.ApplyMessagesToOrderbook(nanoMessages)
```

Comment:
I think you should be iterating over messages rather than by nanoseconds since there'll be many nanoseconds where there will be no corresponding message

Reply:
@chimano will this be possible if we want uniform distance between each point given back

Reply:
Definitely, the state of the orderbook doesn't change between messages so if we have m1 at 97s and m2 at 105s and we want points at a 100s interval, m1 will be applied, once m2 is reached we can see that it's at another multiple, so then the best bid/ask will be stored at 100s and then it'll continue applying the rest of the messages for the other times

Reply:
i did this because there will sometimes be multiple messages at a nanosecond that need to be applied together

Reply:
with some extra logic you can still do that by iterating messages